File: runtime/lib/phases/resolve-pointers.ts
import { existsSync } from 'node:fs';
import {
  type Artifact,
  type Integrity,
  type Manifest,
  type PointerDescriptor,
  type Source,
  artifactApplies,
  fetchWithRetry,
  interpolate,
  isSourcePointer,
  parsePointerDescriptor,
} from '@torba/core';
import type { OsOptions } from '@torba/core';
import { NetworkError } from '../errors';
import { verifyIntegrity } from './verify';

/** Guards against a descriptor whose `source` loops back to a pointer. */
const MAX_POINTER_DEPTH = 5;

export interface PointerResolution {
  /** Manifest with every applicable `pointer` source rewritten in place. */
  manifest: Manifest;
  /**
   * `artifact.path` templates whose on-disk copy is stale (hash no longer
   * matches the descriptor, or the descriptor carries no hash). `scan`
   * must refetch these even though the file exists.
   */
  refetch: Set<string>;
  /** Count of pointer artifacts resolved — for progress reporting. */
  resolved: number;
}

async function fetchDescriptor(url: string): Promise<PointerDescriptor> {
  const res = await fetchWithRetry(url);
  if (!res.ok) {
    const body = await res.text().catch(() => '');
    throw new NetworkError(url, res.status, body.slice(0, 200));
  }
  return parsePointerDescriptor(await res.text());
}

/** Follow a pointer (and any pointer it resolves to) to a concrete source. */
async function follow(
  source: Source,
  vars: Record<string, string>,
): Promise<{ source: Source; integrity?: Integrity; size?: number }> {
  let current = source;
  let integrity: Integrity | undefined;
  let size: number | undefined;
  for (let depth = 0; isSourcePointer(current); depth++) {
    if (depth >= MAX_POINTER_DEPTH) {
      throw new Error(
        `Pointer chain exceeded ${MAX_POINTER_DEPTH} hops at ${current.pointer}`,
      );
    }
    const descriptor = await fetchDescriptor(
      interpolate(current.pointer, vars),
    );
    current = descriptor.source;
    integrity = descriptor.integrity;
    size = descriptor.size;
  }
  return { source: current, integrity, size };
}

/**
 * Resolve every `pointer` source against its live descriptor. Runs before
 * `scan` so the rest of the pipeline only ever sees concrete sources.
 *
 * Non-applicable artifacts (filtered out by this platform's rules) keep
 * their pointer source untouched — no point fetching a descriptor for an
 * artifact `scan` will drop anyway.
 */
export async function resolvePointers(
  manifest: Manifest,
  vars: Record<string, string>,
  platform: OsOptions,
): Promise<PointerResolution> {
  const refetch = new Set<string>();
  let resolved = 0;

  const artifacts = await Promise.all(
    manifest.artifacts.map(async (artifact): Promise<Artifact> => {
      if (!isSourcePointer(artifact.source)) return artifact;
      if (!artifactApplies(artifact, platform)) return artifact;
      resolved++;

      const r = await follow(artifact.source, vars);
      const next: Artifact = {
        ...artifact,
        source: r.source,
        integrity: r.integrity ?? artifact.integrity,
        size: r.size ?? artifact.size,
      };

      // Freshness: keep an existing local copy only if it still matches the
      // descriptor's hash. No hash → the channel is mutable and unverifiable
      // by content, so always refetch to stay current.
      const finalPath = interpolate(next.path, vars);
      if (existsSync(finalPath)) {
        const fresh = next.integrity
          ? await verifyIntegrity(finalPath, next.integrity)
          : false;
        if (!fresh) refetch.add(next.path);
      }
      return next;
    }),
  );

  return { manifest: { ...manifest, artifacts }, refetch, resolved };
}

File: crates/opys-runtime/src/phases/resolve_pointers.rs
use indexmap::IndexMap;
use std::collections::HashSet;
use std::path::Path;
use opys_core::{
    artifact_applies, interpolate, parse_pointer_descriptor, Artifact, Manifest, OsOptions, Source,
};

use crate::errors::InstallError;
use crate::fetch::{fetch_with_retry, RetryOptions};
use crate::phases::verify::verify_integrity;

const MAX_POINTER_DEPTH: usize = 5;

pub struct PointerResolution {
    pub manifest: Manifest,
    pub refetch: HashSet<String>,
    pub resolved: u32,
}

async fn fetch_descriptor(url: &str) -> Result<opys_core::PointerDescriptor, InstallError> {
    let res = fetch_with_retry(reqwest::Method::GET, url, RetryOptions::default())
        .await
        .map_err(|e| InstallError::Network {
            url: url.to_owned(),
            status: 0,
            body: e.to_string(),
        })?;
    let status = res.status();
    if !status.is_success() {
        let body = res.text().await.unwrap_or_default();
        return Err(InstallError::Network {
            url: url.to_owned(),
            status: status.as_u16(),
            body: body.chars().take(200).collect(),
        });
    }
    let text = res.text().await.map_err(|e| InstallError::Network {
        url: url.to_owned(),
        status: 0,
        body: e.to_string(),
    })?;
    parse_pointer_descriptor(&text).map_err(InstallError::Other)
}

async fn follow(
    source: Source,
    vars: &IndexMap<String, String>,
) -> Result<(Source, Option<opys_core::Integrity>, Option<u64>), InstallError> {
    let mut current = source;
    let mut integrity: Option<opys_core::Integrity> = None;
    let mut size: Option<u64> = None;
    let mut depth = 0;
    loop {
        let Source::Pointer { pointer } = &current else {
            break;
        };
        if depth >= MAX_POINTER_DEPTH {
            return Err(InstallError::other(format!(
                "Pointer chain exceeded {MAX_POINTER_DEPTH} hops"
            )));
        }
        let url = interpolate(pointer, vars);
        let descriptor = fetch_descriptor(&url).await?;
        current = descriptor.source;
        integrity = descriptor.integrity;
        size = descriptor.size;
        depth += 1;
    }
    Ok((current, integrity, size))
}

pub async fn resolve_pointers(
    manifest: Manifest,
    vars: &IndexMap<String, String>,
    platform: &OsOptions,
) -> Result<PointerResolution, InstallError> {
    let mut refetch = HashSet::new();
    let mut resolved = 0u32;
    let mut new_artifacts = Vec::with_capacity(manifest.artifacts.len());

    for artifact in manifest.artifacts {
        if !matches!(artifact.source, Source::Pointer { .. })
            || !artifact_applies(&artifact, platform, &[])?
        {
            new_artifacts.push(artifact);
            continue;
        }
        resolved += 1;
        let (source, integ, size) = follow(artifact.source.clone(), vars).await?;
        let next = Artifact {
            source,
            integrity: integ.clone().or_else(|| artifact.integrity.clone()),
            size: size.or(artifact.size),
            ..artifact
        };
        let final_path = interpolate(&next.path, vars);
        if Path::new(&final_path).exists() {
            let fresh = match next.integrity.as_ref() {
                Some(i) => verify_integrity(&final_path, Some(i)).await,
                None => false,
            };
            if !fresh {
                refetch.insert(next.path.clone());
            }
        }
        new_artifacts.push(next);
    }

    let manifest = Manifest {
        artifacts: new_artifacts,
        ..manifest
    };
    Ok(PointerResolution {
        manifest,
        refetch,
        resolved,
    })
}
